File: repos_estudiantes_ALVAREZ_tercera_entrega/AJATA_CANAVIRI_NICOL_MAYRA/Generecidad/Ejercicio3/Python/Catalogo.py
from typing import TypeVar, Generic, List, Optional

T = TypeVar('T')

class Catalogo(Generic[T]):

    def __init__(self):
        self._elementos: List[T] = []

    # a) 
    def agregar_elemento(self, elemento: T):
       
        self._elementos.append(elemento)
        print(f"Elemento agregado al catalogo: {elemento}")

    def buscar_elemento(self, criterio: str) -> Optional[T]:
        
        print(f"Buscando '{criterio}' en el catalogo")
        for elemento in self._elementos:
           
            if hasattr(elemento, 'es_similar') and callable(getattr(elemento, 'es_similar')):
                if elemento.es_similar(criterio):
                    return elemento
        return None

    def mostrar_catalogo(self):
        
        if not self._elementos:
            print("El catalogo esta vacio")
            return

        print("\n--- Contenido Actual del Catalogo ---")
        for elemento in self._elementos:
            print(elemento)
        print("------------------------------------")

File: repos_estudiantes_ALVAREZ_tercera_entrega/AJATA_CANAVIRI_NICOL_MAYRA/Generecidad/Ejercicio3/Java/Catalogo.java
package Generecidad.Ejercicio3.Java;
import java.util.ArrayList;
import java.util.List;


public class Catalogo<T extends Buscarable> {
    private List<T> elementos;

    public Catalogo() {
        this.elementos = new ArrayList<>();
    }

    // a) 

    public void agregarElemento(T elemento) {
        this.elementos.add(elemento);
        System.out.println("Elemento agregado al catalogo: " + elemento);
    }


    public T buscarElemento(String criterio) {
        System.out.println("Buscando '" + criterio + "' en el catalogo");
        for (T elemento : elementos) {
            if (elemento.esSimilar(criterio)) { 
                return elemento;
            }
        }
        return null; 
    }

    public void mostrarCatalogo() {
        if (elementos.isEmpty()) {
            System.out.println("El catalogo esta vacio");
            return;
        }

        System.out.println("\n--- Contenido Actual del Catalogo ---");
        for (T elemento : elementos) {
            System.out.println(elemento);
        }
        System.out.println("------------------------------------");
    }
}
